### PR TITLE
upgrade(jsonschema): jsonschema==3.2.0

### PR DESCRIPTION
**Pull request by BYK**
We need this for Py 3.6+ compatibility. The changelog for the project itself doesn't report any breaking changes for Python side. The main change is making [draft 7](json-schema.org/draft-07) the default which states that it is fully backward-compoatible with draft 6, the current default:

>For the Core and Validation specifications, draft-07 is a relatively minor update. In terms of validation keywords and outcomes, it is fully backwards-compatible with draft-06.

At the end of [Validating Formats](https://python-jsonschema.readthedocs.io/en/stable/validate/#validating-formats), we have the following note:

> Certain checkers will only be available if an appropriate package is available for use. The easiest way to ensure you have what is needed is to install jsonschema using the format or format_nongpl setuptools extra – i.e.
>
>$ pip install jsonschema[format]
>
>which will install all of the below dependencies for all formats.
>
>Or if you want to install MIT-license compatible dependencies only:
>
>$ pip install jsonschema[format_nongpl]
>
>The non-GPL extra is intended to not install any direct dependencies that are GPL (but that of course end-users should do their own verification). At the moment, it supports all the available checkers except for iri and iri-reference.

>*If the following packages are not installed when using a checker that requires it, validation will succeed without throwing an error, as specified by the JSON Schema specification.*

Hence, our failed test on the first try. Glad we had a test for this.

https://app.asana.com/0/0/1161741586199981/f

### CODE REVIEW

File: requirements-base.txt [hunk 1]
```diff
@@ -23,7 +23,7 @@ google-cloud-pubsub==0.35.4
 google-cloud-storage==1.13.3
 googleapis-common-protos==1.6.0
 ipaddress>=1.0.16,<1.1.0 ; python_version < "3.3"
-jsonschema==2.6.0
+jsonschema[format_nongpl]==3.2.0
```
Comment (joshuarli) [inline requirements-base.txt:26]: nice

**Comment by joshuarli** [comment]
> The main change is making draft 7 the default which states that it is fully backward-compoatible with draft 6, the current default

2.6.0 doesn't support Draft 6 at all, it's Draft 4 which is default.

**Comment by joshuarli** [comment]
We use Draft4Validator in a lot of places, so that's fine. But the calls to `jsonschema.validate` should use Draft 4 as well since that's the default in 2.6.0: https://python-jsonschema.readthedocs.io/en/latest/validate/#jsonschema.validate

It is also entirely possible if all tests pass that we don't have enough coverage, or we're compatible with Draft 6/7. Here's 4 -> 6: https://json-schema.org/draft-06/json-schema-release-notes.html

**Comment by wedamija** [comment]
> We use Draft4Validator in a lot of places, so that's fine. But the calls to `jsonschema.validate` should use Draft 4 as well since that's the default in 2.6.0: https://python-jsonschema.readthedocs.io/en/latest/validate/#jsonschema.validate
> 
> It is also entirely possible if all tests pass that we don't have enough coverage, or we're compatible with Draft 6/7. Here's 4 -> 6: https://json-schema.org/draft-06/json-schema-release-notes.html

I did a quick scan of our schemas and they all look compatible based on that spec. Probably worth confirming

**Comment by BYK** [comment]
Yeah I'm doing more auditing right now. Good catch with Draft4 @joshuarli.

Right now the builds are failing as the `idna` requirement from our older `requests` version is incompatible with the `idna` version installed by `jsonschema[format_nongpl]`.

We have two options:

1. Upgrade `requests` to v2.23.0, which seems harmless: https://github.com/psf/requests/compare/v2.20.0...v2.23.0
2. Do a bare install of `jsonschema` and install `rfc3986-validator` and `rfc3339-validator` manually with notes

Option 1 is nice as it makes using `jsonschema` usage less of a mine-field and comes with a more recent version of `requests`. Option 2 is nice as it avoids installing unnecessary extra dependencies like `webcolors` and it also requires the minimum changes, avoiding a potentially dangerous `requests` upgrade.

Any thoughts @wedamija @joshuarli?

**Comment by BYK** [comment]
Regarding Draft 4 -> Draft 7 transition, the only thing that _may_ affect us is the slight definition change on `"type": "integer"`: https://json-schema.org/draft-06/json-schema-release-notes.html#backwards-incompatible-changes

This is only used for Snuba so it doesn't pose any risk IMO. We can follow this patch up with an upgrade to Draft 7 in all the places.

**Comment by joshuarli** [comment]
> Yeah I'm doing more auditing right now. Good catch with Draft4 @joshuarli.
> 
> Right now the builds are failing as the `idna` requirement from our older `requests` version is incompatible with the `idna` version installed by `jsonschema[format_nongpl]`.
> 
> We have two options:
> 
>     1. Upgrade `requests` to v2.23.0, which seems harmless: [psf/requests@v2.20.0...v2.23.0](https://github.com/psf/requests/compare/v2.20.0...v2.23.0)
> 
>     2. Do a bare install of `jsonschema` and install `rfc3986-validator` and `rfc3339-validator` manually with notes
> 
> 
> Option 1 is nice as it makes using `jsonschema` usage less of a mine-field and comes with a more recent version of `requests`. Option 2 is nice as it avoids installing unnecessary extra dependencies like `webcolors` and it also requires the minimum changes, avoiding a potentially dangerous `requests` upgrade.
> 
> Any thoughts @wedamija @joshuarli?

I would prefer Option 2 since it's least effort and least risk. There's other considerations with bumping requests like possibly needing to also bump urllib3 in getsentry, and I'm not sure how it'll affect usage of `responses` in our test suite. 

Option 1 would be nice to have though, and I can see it being done in the future to address CVEs for example, but it's not a blocker for strictly py3.

**Comment by wedamija** [comment]
> > Yeah I'm doing more auditing right now. Good catch with Draft4 @joshuarli.
> > Right now the builds are failing as the `idna` requirement from our older `requests` version is incompatible with the `idna` version installed by `jsonschema[format_nongpl]`.
> > We have two options:
> > ```
> > 1. Upgrade `requests` to v2.23.0, which seems harmless: [psf/requests@v2.20.0...v2.23.0](https://github.com/psf/requests/compare/v2.20.0...v2.23.0)
> > 
> > 2. Do a bare install of `jsonschema` and install `rfc3986-validator` and `rfc3339-validator` manually with notes
> > ```
> > 
> > 
> > Option 1 is nice as it makes using `jsonschema` usage less of a mine-field and comes with a more recent version of `requests`. Option 2 is nice as it avoids installing unnecessary extra dependencies like `webcolors` and it also requires the minimum changes, avoiding a potentially dangerous `requests` upgrade.
> > Any thoughts @wedamija @joshuarli?
> 
> I would prefer Option 2 since it's least effort and least risk. There's other considerations with bumping requests like possibly needing to also bump urllib3 in getsentry, and I'm not sure how it'll affect usage of `responses` in our test suite.
> 
> Option 1 would be nice to have though, and I can see it being done in the future to address CVEs for example, but it's not a blocker for strictly py3.

I agree with Option 2 as well. I think that generally we should only install the dependencies we specifically need, so specifying those validator libraries manually seems best. I also agree with reducing risk by not bumping `requests` at the moment. We're already adding a  bunch of potential instability in general with this project, so best to avoid more when we can.

**Review by wedamija — APPROVED** [review]

**Comment by joshuarli** [comment]
Wait so, we're going to be switching to Draft 7 then as default validator? I was expecting some commits to change `jsonschema.validate` to Draft 4 and then re-request review.

**Comment by BYK** [comment]
>Wait so, we're going to be switching to Draft 7 then as default validator? I was expecting some commits to change jsonschema.validate to Draft 4 and then re-request review.

Yes as we only have one place that _may_ have a bw-incompatibility issue which is the snuba version field (now accepts `1.0` as an integer, which is fine). The effort to find and change all the places for raw validate to Draft4 is not worth it.

I'll be following up with a patch that moves everything over to Draft7.

**Comment by wedamija** [comment]
> > Wait so, we're going to be switching to Draft 7 then as default validator? I was expecting some commits to change jsonschema.validate to Draft 4 and then re-request review.
> 
> Yes as we only have one place that _may_ have a bw-incompatibility issue which is the snuba version field (now accepts `1.0` as an integer, which is fine). The effort to find and change all the places for raw validate to Draft4 is not worth it.
> 
> I'll be following up with a patch that moves everything over to Draft7.

Yeah that's fine, the version is something we specify in snuba. If they passed it as a decimal it might actually fail, but as long as it doesn't convert and in to 1.0 then we're fine.

**Pull request merged**